software/base/src/main/java/brooklyn/entity/basic/AbstractSoftwareProcessWinRmDriver.java

```diff
@@ -0,0 +1,192 @@
+        if (createParentDir) {
+            createDirectory(getDirectory(target), "Creating resource directory");
+        }
+        return copyTo(new File(source), new File(target));
+    }
+
+    @Override
+    public int copyResource(Map<Object, Object> sshFlags, InputStream source, String target, boolean createParentDir) {
+        if (createParentDir) {
+            createDirectory(getDirectory(target), "Creating resource directory");
+        }
+        return copyTo(source, new File(target));
+    }
+
+    @Override
+    protected void createDirectory(String directoryName, String summaryForLogging) {
+        getLocation().executePsScript("New-Item -path \"" + directoryName + "\" -type directory -ErrorAction SilentlyContinue");
+    }
+
+    protected WinRmToolResponse executeCommand(ConfigKey<String> regularCommandKey, ConfigKey<String> powershellCommandKey, boolean allowNoOp) {
+        String regularCommand = getEntity().getConfig(regularCommandKey);
+        String powershellCommand = getEntity().getConfig(powershellCommandKey);
+        if (Strings.isNullOrEmpty(regularCommand) && Strings.isNullOrEmpty(powershellCommand)) {
+            if (allowNoOp) {
+                return new WinRmToolResponse("", "", 0);
+            } else {
+                throw new IllegalStateException(String.format("Exactly one of %s or %s must be set", regularCommandKey.getName(), powershellCommandKey.getName()));
+            }
+        } else if (!Strings.isNullOrEmpty(regularCommand) && !Strings.isNullOrEmpty(powershellCommand)) {
+            throw new IllegalStateException(String.format("%s and %s cannot both be set", regularCommandKey.getName(), powershellCommandKey.getName()));
+        }
+
+        if (Strings.isNullOrEmpty(regularCommand)) {
+            return getLocation().executePsScript(ImmutableList.of(powershellCommand));
+        } else {
+            return getLocation().executeScript(ImmutableList.of(regularCommand));
+        }
+    }
+
+    public int execute(List<String> script) {
+        return getLocation().executeScript(script).getStatusCode();
+    }
+
+    public int executePowerShellNoRetry(List<String> psScript) {
+        return getLocation().executePsScriptNoRetry(psScript).getStatusCode();
+    }
+
+    public int executePowerShell(List<String> psScript) {
+        return getLocation().executePsScript(psScript).getStatusCode();
+    }
+
+    public int copyTo(File source, File destination) {
+        return getLocation().copyTo(source, destination);
+    }
+
+    public int copyTo(InputStream source, File destination) {
+        return getLocation().copyTo(source, destination);
+    }
+
+    public void rebootAndWait() {
+        try {
+            executePowerShellNoRetry(ImmutableList.of("Restart-Computer -Force"));
+        } catch (PyException e) {
+            // Restarting the computer will cause the command to fail; ignore the exception and continue
```

Comment:
Can we be more specific - e.g. what exception do you expect. And log if not exactly right. There's a risk we'll pretend the VM is rebooting when actually we got back an error about permissions or something like that.